src/compiler/checker.ts

```diff
@@ -19673,7 +19673,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         if (isJSConstructor(container) && isNodeDescendantOf(node, container.body)) {
             return getDeclaredTypeOfClassOrInterface(getSymbolOfDeclaration(container)).thisType!;
         }
-        error(node, Diagnostics.A_this_type_is_available_only_in_a_non_static_member_of_a_class_or_interface);
+        if (!(node.parent.kind === SyntaxKind.TypePredicate && node === (node.parent as TypePredicateNode).parameterName)) {
+            error(node, Diagnostics.A_this_type_is_available_only_in_a_non_static_member_of_a_class_or_interface);
+        }
         return errorType;

```

Comment:
The change here avoids the error message, but it's a little weird that in this case, the type would be `errorType`.

You can see from the current behavior that if you hover over `this`, you'll see `any`, which is weird if `this` has an explicit type in a case like:

```ts
let x = {
    isYes(this: typeof x): this is Object {
        //                 ^?
        return true;
    }
}
```

Ideally none of this would go through `getTypeFromThisTypeNode` in the first place.

Reply:
Yeah, it looks like the inbound call from `getTypeFromThisTypeNode` is bogus.